Building newer versions of Firebase for iOS faills

Newer versions of Firebase require module support. Adding -fmodules and -fcxx-modules results in error "cyclic dependency in module 'Darwin': Darwin -> std -> Darwin"

```
clang++ -c -arch arm64 -target arm-apple-darwin19 -m64 -O2 -g -stdlib=libc++ -miphoneos-version-min=11.0 -isysroot /Users/bjornritzl/extender/platformsdk/iPhoneOS16.2.sdk -nostdinc++ -isystem /Users/bjornritzl/extender/platformsdk/iPhoneOS16.2.sdk/usr/include/c++/v1  -DDLIB_LOG_DOMAIN="FIREBASECORE" -DDDF_EXPOSE_DESCRIPTORS -DDM_PLATFORM_IOS -DDM_DEBUG -DFirebase_VERSION=10.12.0 -DCOCOAPODS=1  -fno-autolink -fobjc-arc --language=objective-c -fmodules -fmodule-name=FirebaseCore  -Iupload/ -Iupload/firebase/include/ -ICocoaPodsService/Pods/FirebaseCore/FirebaseCore/Sources/Public/FirebaseCore/ -ICocoaPodsService/Pods/FirebaseCore/FirebaseCore/Sources/Public/ -ICocoaPodsService/Pods/FirebaseCore/FirebaseCore/Sources/ -ICocoaPodsService/Pods/FirebaseCore/FirebaseCore/ -ICocoaPodsService/Pods/FirebaseCore/FirebaseCore/Extension/ -ICocoaPodsService/Pods/FirebaseCore/ -ICocoaPodsService/Pods/FirebaseCoreInternal/ -ICocoaPodsService/Pods/FirebaseInstallations/FirebaseInstallations/Source/Library/InstallationsIDController/ -ICocoaPodsService/Pods/FirebaseInstallations/FirebaseInstallations/Source/Library/ -ICocoaPodsService/Pods/FirebaseInstallations/FirebaseInstallations/Source/Library/Public/FirebaseInstallations/ -ICocoaPodsService/Pods/FirebaseInstallations/FirebaseInstallations/Source/Library/Public/ -ICocoaPodsService/Pods/FirebaseInstallations/FirebaseInstallations/Source/Library/Private/ -ICocoaPodsService/Pods/FirebaseInstallations/FirebaseInstallations/Source/Library/InstallationsAPI/ -ICocoaPodsService/Pods/FirebaseInstallations/FirebaseInstallations/Source/Library/Errors/ -ICocoaPodsService/Pods/FirebaseInstallations/FirebaseInstallations/Source/Library/IIDMigration/ -ICocoaPodsService/Pods/FirebaseInstallations/FirebaseInstallations/Source/Library/InstallationsStore/ -ICocoaPodsService/Pods/FirebaseInstallations/FirebaseInstallations/Source/ -ICocoaPodsService/Pods/FirebaseInstallations/FirebaseCore/Extension/ -ICocoaPodsService/Pods/FirebaseInstallations/FirebaseCore/ -ICocoaPodsService/Pods/FirebaseInstallations/ -ICocoaPodsService/Pods/GoogleUtilities/GoogleUtilities/Environment/Public/GoogleUtilities/ -ICocoaPodsService/Pods/GoogleUtilities/GoogleUtilities/Environment/Public/ -ICocoaPodsService/Pods/GoogleUtilities/ -ICocoaPodsService/Pods/GoogleUtilities/GoogleUtilities/Logger/Public/GoogleUtilities/ -ICocoaPodsService/Pods/GoogleUtilities/GoogleUtilities/Logger/Public/ -ICocoaPodsService/Pods/GoogleUtilities/ -ICocoaPodsService/Pods/GoogleUtilities/GoogleUtilities/NSData+zlib/Public/GoogleUtilities/ -ICocoaPodsService/Pods/GoogleUtilities/GoogleUtilities/NSData+zlib/Public/ -ICocoaPodsService/Pods/GoogleUtilities/ -ICocoaPodsService/Pods/GoogleUtilities/GoogleUtilities/UserDefaults/Public/GoogleUtilities/ -ICocoaPodsService/Pods/GoogleUtilities/GoogleUtilities/UserDefaults/Public/ -ICocoaPodsService/Pods/GoogleUtilities/ -ICocoaPodsService/Pods/PromisesObjC/Sources/FBLPromises/include/ -ICocoaPodsService/Pods/PromisesObjC/Sources/FBLPromises/ -ICocoaPodsService/Pods/PromisesObjC/ -ICocoaPodsService/frameworks/headers/arm64-ios/  -F/Users/bjornritzl/Downloads/EXTENDER/job1/CocoaPodsService/frameworks/lib/arm64-ios  -I/Users/bjornritzl/projects/defold/tmp/dynamo_home//include -I/Users/bjornritzl/projects/defold/tmp/dynamo_home//sdk/include -I/Users/bjornritzl/projects/defold/tmp/dynamo_home//ext/include   CocoaPodsService/Pods/FirebaseCore/FirebaseCore/Sources/FIRDependency.m -obuild/FIRDependency.m_7.o
While building module 'Foundation' imported from CocoaPodsService/Pods/FirebaseCore/FirebaseCore/Extension/FIRComponentType.h:17:
While building module 'CoreFoundation' imported from /Users/bjornritzl/extender/platformsdk/iPhoneOS16.2.sdk/System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:6:
While building module 'Darwin' imported from /Users/bjornritzl/extender/platformsdk/iPhoneOS16.2.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CoreFoundation.h:16:
While building module 'std' imported from /Users/bjornritzl/extender/platformsdk/iPhoneOS16.2.sdk/usr/include/tgmath.h:31:
In file included from <module-includes>:1:
/Users/bjornritzl/extender/platformsdk/iPhoneOS16.2.sdk/usr/include/c++/v1/ctype.h:38:15: fatal error: cyclic dependency in module 'Darwin': Darwin -> std -> Darwin
#include_next <ctype.h>
```

--- a/README_RELEASE.md
+++ b/README_RELEASE.md
@@ -23,7 +23,6 @@ This will create a new task definition on AWS ECS and update the service to run 
 Note, we currently don't have any macOS specific tests, so you may skip the tests
 
   1. Build (without tests): `./server/scripts/build-standalone.sh -xtest`
-  1. Run: `./server/scripts/build-standalone.sh -xtest`
   1. Run `./server/scripts/publish-darwin-stage.sh`
 
 This will upload the packages and jar files to the macOS machine on AWS, and restart the extender service.

--- a/server/src/main/java/com/defold/extender/Extender.java
+++ b/server/src/main/java/com/defold/extender/Extender.java
@@ -760,6 +760,14 @@ class Extender {
         Map<String, Object> mergedContextWithPodsForObjC = ExtenderUtil.mergeContexts(trimmedContext, podContextObjC);
         Map<String, Object> mergedContextWithPodsForObjCpp = ExtenderUtil.mergeContexts(trimmedContext, podContextObjCpp);
 
+        // remove systemIncludes from objc and objc++
+        // this is a bit crude but cocoapod builds do not provide any -isystem option and
+        // it seems like the "{{env.SYSROOT}}/usr/include/c++/v1" set in build.yml is
+        // causing problems when building objc code with -fmodules enabled
+        // see https://github.com/defold/extender/issues/308
+        mergedContextWithPodsForObjC.put("systemIncludes", new ArrayList<String>());
+        mergedContextWithPodsForObjCpp.put("systemIncludes", new ArrayList<String>());
+
         for (File src : pod.sourceFiles) {
             String extension = FilenameUtils.getExtension(src.getAbsolutePath());
             if (extension.equals("swift")) {

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -32,8 +32,10 @@ import java.util.List;
 import java.util.Map;
 import java.util.HashMap;
 import java.util.Set;
+import java.util.HashSet;
 import java.util.LinkedHashSet;
 import java.util.Iterator;
+import java.util.Collections;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.lang.StringBuilder;
@@ -224,6 +226,7 @@ public class CocoaPodsService {
         public PodSpec parentSpec = null;
         public List<String> defaultSubspecs = new ArrayList<>();
         public List<PodSpec> subspecs = new ArrayList<>();
+        public List<String> dependencies = new ArrayList<>();
 
         public PlatformAndLanguageSet flags = new PlatformAndLanguageSet();
         public PlatformSet defines = new PlatformSet();
@@ -234,6 +237,15 @@ public class CocoaPodsService {
         public PlatformSet frameworks = new PlatformSet();
         public PlatformSet libraries = new PlatformSet();
         public File dir;
+
+        public PodSpec getSubspec(String name) {
+            for (PodSpec spec : subspecs) {
+                if (spec.name.equals(name)) {
+                    return spec;
+                }
+            }
+            return  null;
+        }
 
         @Override
         public String toString() {
@@ -250,12 +262,13 @@ public class CocoaPodsService {
             sb.append("  frameworks: " + frameworks + "\n");
             sb.append("  vendored_frameworks: " + vendoredFrameworks + "\n");
             sb.append("  libraries: " + libraries + "\n");
+            sb.append("  dependencies: " + dependencies + "\n");
             sb.append("  parentSpec: " + ((parentSpec != null) ? parentSpec.name : "null") + "\n");
+            sb.append("  default_subspecs: " + defaultSubspecs + "\n");
             for (PodSpec sub : subspecs) {
                 sb.append("  subspec: " + sub.name + "\n");
             }
             return sb.toString();
-   
         }
     }
 
@@ -844,6 +857,14 @@ public class CocoaPodsService {
             }
         }
 
+        // parse dependencies
+        Map<String, List<Object>> dependencies = (Map<String, List<Object>>)specJson.get("dependencies");
+        if (dependencies != null) {
+            for (String dependency : dependencies.keySet()) {
+                spec.dependencies.add(dependency);
+            }
+        }
+
         // find source and header files
         // https://guides.cocoapods.org/syntax/podspec.html#source_files
         List<String> sourceFilePatterns = new ArrayList<>();
@@ -876,6 +897,55 @@ public class CocoaPodsService {
         PodSpec spec = createPodSpec(parseJson(specJson), null, podsDir, jobEnvContext);
         spec.originalJSON = specJson;
         return spec;
+    }
+
+    // 'GoogleUtilities/Environment (7.10.0)'  -> 'GoogleUtilities/Environment' -> ['GoogleUtilities', 'Environment']
+    private String[] splitPodname(String pod) {
+        return pod.replaceFirst(" \\(.*\\)", "").split("/");
+    }
+
+    /**
+     * Get pod spec based on a pod name with optional sub pod (GoogleUtilities/Environment)
+     * @param pods Map of pod names to pod specs
+     * @param podname The pod to find
+     * @return The pod spec
+     */
+    private PodSpec getPod(Map<String, PodSpec> pods, String podname) throws ExtenderException {
+        // 'GoogleUtilities/Environment (7.10.0)'  -> 'GoogleUtilities/Environment' -> ['GoogleUtilities', 'Environment']
+        String podnameparts[] = splitPodname(podname);
+        // 'GoogleUtilities'
+        String mainpodname = podnameparts[0];
+        PodSpec current = pods.get(mainpodname);
+        if (podnameparts.length > 1) {
+            for (int i = 1; i < podnameparts.length; i++) {
+                String subspecname = podnameparts[i];
+                PodSpec subspec = current.getSubspec(subspecname);
+                if (subspec == null) {
+                    throw new ExtenderException("Unable to find subspec '" + subspecname + "' in pod '" + current.name + "'");
+                }
+                current = subspec;
+            }
+        }
+        return current;
+    }
+
+    /**
+     * Get a sorted set of pod specs with dependencies from a list of pod names.
+     * This function will recursively go through all pods and add their
+     * dependencies to the final set of pods. The pod specs will be added to the
+     * set such that the dependencies of a pod are added before their pod itself.
+     * @param specsMap Map with pod specs to search for pods
+     * @param podnames Names of the pods to get
+     * @return Set of pod specs
+     */
+    private LinkedHashSet<PodSpec> getSpecsAndDependencies(Map<String, PodSpec> specsMap, List<String> podnames) throws ExtenderException {
+        LinkedHashSet<PodSpec> specs = new LinkedHashSet<>();
+        for (String podname : podnames) {
+            PodSpec spec = getPod(specsMap, podname);
+            specs.addAll(getSpecsAndDependencies(specsMap, spec.dependencies));
+            specs.add(spec);
+        }
+        return specs;
     }
 
     /**
@@ -913,28 +983,28 @@ public class CocoaPodsService {
             - GoogleAppMeasurement (= 8.13.0)
         */
         List<String> lines = Files.readAllLines(podFileLock.toPath());
-        Set<String> pods = new LinkedHashSet<>();
+        Set<String> podnames = new LinkedHashSet<>();
         while (!lines.isEmpty()) if (lines.remove(0).startsWith("PODS:")) break;
         while (!lines.isEmpty()) {
             String line = lines.remove(0);
             if (line.trim().isEmpty()) break;
             if (line.startsWith("  -")) {
                 // '  - "GoogleUtilities/Environment (7.10.0)":'   ->   'GoogleUtilities/Environment (7.10.0)'
-                String pod = line.trim().replace("- ", "").replace(":", "").replace("\"","");
-                pods.add(pod);
+                String podname = line.trim().replace("- ", "").replace(":", "").replace("\"","");
+                podnames.add(podname);
             }
         }
 
+        // get all the pod specs and store them in a map, keyed on pod name
         File podsDir = new File(workingDir, "Pods");
-        List<PodSpec> specs = new ArrayList<>();
         Map<String, PodSpec> specsMap = new HashMap<>();
-        for (String pod : pods) {
+        for (String podname : podnames) {
             // 'GoogleUtilities/Environment (7.10.0)'  -> 'GoogleUtilities/Environment' -> ['GoogleUtilities', 'Environment']
-            String podnameparts[] = pod.replaceFirst(" \\(.*\\)", "").split("/");
+            String podnameparts[] = splitPodname(podname);
             // 'GoogleUtilities'
             String mainpodname = podnameparts[0];
             // 'GoogleUtilities/Environment (7.10.0)'  -> '7.10.0'
-            String podversion = pod.replaceFirst(".*\\(", "").replace(")", "");
+            String podversion = podname.replaceFirst(".*\\(", "").replace(")", "");
             if (!specsMap.containsKey(mainpodname)) {
                 String cmd = "pod spec cat --regex ^" + mainpodname + "$ --version=" + podversion;
                 String specJson = execCommand(cmd).replace(cmd, "");
@@ -951,26 +1021,12 @@ public class CocoaPodsService {
 
                 specsMap.put(mainpodname, createPodSpec(specJson, podsDir, jobEnvContext));
             }
-
-            PodSpec mainpod = specsMap.get(mainpodname);
-            if (podnameparts.length == 1) {
-                specs.add(0, mainpod);
-            }
-            else {
-                PodSpec current = mainpod;
-                for (int i = 1; i < podnameparts.length; i++) {
-                    String subspecname = podnameparts[i];
-                    for (PodSpec subspec : current.subspecs) {
-                        if (subspec.name.equals(subspecname)) {
-                            specs.add(0, subspec);
-                            current = subspec;
-                            break;
-                        }
-                    }
-                }
-            }
         }
 
+        // build list of specs and their dependencies (dependencies first)
+        List<String> reversePodnames = new ArrayList<>(podnames);
+        Collections.reverse(reversePodnames);
+        List<PodSpec> specs = new ArrayList<PodSpec>(getSpecsAndDependencies(specsMap, reversePodnames));
         LOGGER.info("Installed pods");
         return specs;
     }

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -591,6 +591,12 @@ public class CocoaPodsService {
         if (hasString(config, "OTHER_LDFLAGS")) {
             linkflags.addAll(getAsSplitString(config, "OTHER_LDFLAGS"));
         }
+        // compiler flags for c and objc files
+        // https://xcodebuildsettings.com/#other_cflags
+        if (hasString(config, "OTHER_CFLAGS")) {
+            flags.c.addAll(getAsSplitString(config, "OTHER_CFLAGS"));
+            flags.objc.addAll(getAsSplitString(config, "OTHER_CFLAGS"));
+        }
         // compiler flags
         if (hasString(config, "CLANG_CXX_LANGUAGE_STANDARD")) {
             String cppStandard = getAsString(config, "CLANG_CXX_LANGUAGE_STANDARD", "compiler-default");

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -752,6 +752,8 @@ public class CocoaPodsService {
         // https://guides.cocoapods.org/syntax/podspec.html#compiler_flags
         spec.flags.ios.addAll(getAsSplitString(specJson, "compiler_flags"));
         spec.flags.osx.addAll(getAsSplitString(specJson, "compiler_flags"));
+        if (ios != null) spec.flags.ios.addAll(getAsJSONArray(ios, "compiler_flags"));
+        if (osx != null) spec.flags.osx.addAll(getAsJSONArray(osx, "compiler_flags"));
         spec.flags.ios.c.add("--language=c");
         spec.flags.osx.c.add("--language=c");
         spec.flags.ios.cpp.add("--language=c++");
@@ -760,8 +762,16 @@ public class CocoaPodsService {
         spec.flags.osx.objc.add("--language=objective-c");
         spec.flags.ios.objcpp.add("--language=objective-c++");
         spec.flags.osx.objcpp.add("--language=objective-c++");
-        if (ios != null) spec.flags.ios.addAll(getAsJSONArray(ios, "compiler_flags"));
-        if (osx != null) spec.flags.osx.addAll(getAsJSONArray(osx, "compiler_flags"));
+        // CocoaPods sets CLANG_ENABLE_MODULES when creating an XCode project
+        // https://xcodebuildsettings.com/#clang_enable_modules
+        spec.flags.ios.objc.add("-fmodules");
+        spec.flags.osx.objc.add("-fmodules");
+        spec.flags.ios.objcpp.add("-fcxx-modules");
+        spec.flags.osx.objcpp.add("-fcxx-modules");
+        spec.flags.ios.objc.add("-fmodule-name=" + spec.name);
+        spec.flags.osx.objc.add("-fmodule-name=" + spec.name);
+        spec.flags.ios.objcpp.add("-fmodule-name=" + spec.name);
+        spec.flags.osx.objcpp.add("-fmodule-name=" + spec.name);
 
         // resources
         // https://guides.cocoapods.org/syntax/podspec.html#resources

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -588,6 +588,7 @@ public class CocoaPodsService {
             defines.addAll(getAsSplitString(config, "GCC_PREPROCESSOR_DEFINITIONS"));
         }
         // linker flags
+        // https://xcodebuildsettings.com/#other_ldflags
         if (hasString(config, "OTHER_LDFLAGS")) {
             linkflags.addAll(getAsSplitString(config, "OTHER_LDFLAGS"));
         }

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -943,7 +943,7 @@ public class CocoaPodsService {
 
             PodSpec mainpod = specsMap.get(mainpodname);
             if (podnameparts.length == 1) {
-                specs.add(mainpod);
+                specs.add(0, mainpod);
             }
             else {
                 PodSpec current = mainpod;
@@ -951,7 +951,7 @@ public class CocoaPodsService {
                     String subspecname = podnameparts[i];
                     for (PodSpec subspec : current.subspecs) {
                         if (subspec.name.equals(subspecname)) {
-                            specs.add(subspec);
+                            specs.add(0, subspec);
                             current = subspec;
                             break;
                         }

